src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,106 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elements;
+    private int size;
+
+    ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfArrayFull();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        growIfArrayFull();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        growIfArrayFull();
+        for (int i = 0; i < list.size(); i++) {
+            elements[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
 
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removeElement = elements[index];
+        for (int i = index; i < elements.length - 1; i++) {
+            elements[i] = elements[i + 1];
+        }
+        size--;
+        return removeElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T deletedElement;
+        for (int i = 0; i < size; i++) {
+            if ((elements[i] != null
+                    && elements[i].equals(element))
+                    || elements[i] == element) {
```

Comment:
```suggestion
            if ((elements[i] != null && elements[i].equals(element)) || elements[i] == element) {
```

And also you can move those check into a separate method.